File: vertx-lang-scala-stack/vertx-service-discovery/src/main/resources/vertx-service-discovery-js/event_bus_service.js
/*
 * Copyright 2014 Red Hat, Inc.
 *
 * Red Hat licenses this file to you under the Apache License, version 2.0
 * (the "License"); you may not use this file except in compliance with the
 * License.  You may obtain a copy of the License at:
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

/** @module vertx-service-discovery-js/event_bus_service */
var utils = require('vertx-js/util/utils');
var ServiceDiscovery = require('vertx-service-discovery-js/service_discovery');

var io = Packages.io;
var JsonObject = io.vertx.core.json.JsonObject;
var JEventBusService = Java.type('io.vertx.servicediscovery.types.EventBusService');
var Record = Java.type('io.vertx.servicediscovery.Record');

/**

 @class
*/
var EventBusService = function(j_val) {

  var j_eventBusService = j_val;
  var that = this;

  // A reference to the underlying Java delegate
  // NOTE! This is an internal API and must not be used in user code.
  // If you rely on this property your code is likely to break if we change it / remove it without warning.
  this._jdel = j_eventBusService;
};

EventBusService._jclass = utils.getJavaClass("io.vertx.servicediscovery.types.EventBusService");
EventBusService._jtype = {
  accept: function(obj) {
    return EventBusService._jclass.isInstance(obj._jdel);
  },
  wrap: function(jdel) {
    var obj = Object.create(EventBusService.prototype, {});
    EventBusService.apply(obj, arguments);
    return obj;
  },
  unwrap: function(obj) {
    return obj._jdel;
  }
};
EventBusService._create = function(jdel) {
  var obj = Object.create(EventBusService.prototype, {});
  EventBusService.apply(obj, arguments);
  return obj;
}
/**
 Sugar method to creates a record for this type.
 <p>
 The java interface is added to the metadata in the `service.interface` key.

 @memberof module:vertx-service-discovery-js/event_bus_service
 @param name {string} the name of the service. 
 @param address {string} the event bus address on which the service available 
 @param itf {string} the Java interface (name) 
 @param metadata {Object} the metadata 
 @return {Object} the created record
 */
EventBusService.createRecord = function() {
  var __args = arguments;
  if (__args.length === 3 && typeof __args[0] === 'string' && typeof __args[1] === 'string' && typeof __args[2] === 'string') {
    return utils.convReturnDataObject(JEventBusService["createRecord(java.lang.String,java.lang.String,java.lang.String)"](__args[0], __args[1], __args[2]));
  }else if (__args.length === 4 && typeof __args[0] === 'string' && typeof __args[1] === 'string' && typeof __args[2] === 'string' && (typeof __args[3] === 'object' && __args[3] != null)) {
    return utils.convReturnDataObject(JEventBusService["createRecord(java.lang.String,java.lang.String,java.lang.String,io.vertx.core.json.JsonObject)"](__args[0], __args[1], __args[2], utils.convParamJsonObject(__args[3])));
  } else throw new TypeError('function invoked with invalid arguments');
};

/**
 Lookup for a service record and if found, retrieve it and return the service object (used to consume the service).
 This is a convenient method to avoid explicit lookup and then retrieval of the service. This method requires to
 have the <code>clientClass</code> set with the expected set of client. This is important for usages not using Java so
 you can pass the expected type.

 @memberof module:vertx-service-discovery-js/event_bus_service
 @param discovery {ServiceDiscovery} the service discovery 
 @param filter {todo} the filter 
 @param clientClass {todo} the client class 
 @param resultHandler {function} the result handler 
 @return {Object} <code>null</code> - do not use
 */
EventBusService.getServiceProxy = function(discovery, filter, clientClass, resultHandler) {
  var __args = arguments;
  if (__args.length === 4 && typeof __args[0] === 'object' && __args[0]._jdel && typeof __args[1] === 'function' && typeof __args[2] === 'function' && typeof __args[3] === 'function') {
    return utils.get_jtype(__args[2]).wrap(JEventBusService["getServiceProxy(io.vertx.servicediscovery.ServiceDiscovery,java.util.function.Function,java.lang.Class,io.vertx.core.Handler)"](discovery._jdel, function(jVal) {
    var jRet = filter(utils.convReturnDataObject(jVal));
    return jRet;
  }, utils.get_jclass(clientClass), function(ar) {
    if (ar.succeeded()) {
      resultHandler(utils.get_jtype(__args[2]).wrap(ar.result()), null);
    } else {
      resultHandler(null, ar.cause());
    }
  }));
  } else throw new TypeError('function invoked with invalid arguments');
};

/**
 Lookup for a service record and if found, retrieve it and return the service object (used to consume the service).
 This is a convenient method to avoid explicit lookup and then retrieval of the service. This method requires to
 have the <code>clientClass</code> set with the expected set of client. This is important for usages not using Java so
 you can pass the expected type.

 @memberof module:vertx-service-discovery-js/event_bus_service
 @param discovery {ServiceDiscovery} the service discovery 
 @param filter {Object} the filter as json object 
 @param clientClass {todo} the client class 
 @param resultHandler {function} the result handler 
 @return {Object} <code>null</code> - do not use
 */
EventBusService.getServiceProxyWithJsonFilter = function(discovery, filter, clientClass, resultHandler) {
  var __args = arguments;
  if (__args.length === 4 && typeof __args[0] === 'object' && __args[0]._jdel && (typeof __args[1] === 'object' && __args[1] != null) && typeof __args[2] === 'function' && typeof __args[3] === 'function') {
    return utils.get_jtype(__args[2]).wrap(JEventBusService["getServiceProxyWithJsonFilter(io.vertx.servicediscovery.ServiceDiscovery,io.vertx.core.json.JsonObject,java.lang.Class,io.vertx.core.Handler)"](discovery._jdel, utils.convParamJsonObject(filter), utils.get_jclass(clientClass), function(ar) {
    if (ar.succeeded()) {
      resultHandler(utils.get_jtype(__args[2]).wrap(ar.result()), null);
    } else {
      resultHandler(null, ar.cause());
    }
  }));
  } else throw new TypeError('function invoked with invalid arguments');
};

module.exports = EventBusService;

File: vertx-lang-scala-stack/vertx-service-discovery/vertx-service-discovery/src/main/scala/io/vertx/scala/servicediscovery/types/EventBusService.scala
/*
 * Copyright 2014 Red Hat, Inc.
 *
 * Red Hat licenses this file to you under the Apache License, version 2.0
 * (the "License"); you may not use this file except in compliance with the
 * License.  You may obtain a copy of the License at:
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

package io.vertx.scala.servicediscovery.types

import io.vertx.lang.scala.AsyncResultWrapper
import io.vertx.servicediscovery.{Record => JRecord}
import io.vertx.servicediscovery.{ServiceDiscovery => JServiceDiscovery}
import scala.reflect.runtime.universe._
import io.vertx.scala.servicediscovery.ServiceDiscovery
import io.vertx.core.json.JsonObject
import io.vertx.core.AsyncResult
import io.vertx.core.Handler
import io.vertx.servicediscovery.types.{EventBusService => JEventBusService}
import io.vertx.scala.servicediscovery.Record
import io.vertx.lang.scala.HandlerOps._
import io.vertx.lang.scala.Converter._

/**
  *  for event bus services (service proxies).
  * Consumers receive a service proxy to use the service.
  */

class EventBusService(private val _asJava: Object) {
  def asJava = _asJava






}

object EventBusService {
  def apply(asJava: JEventBusService) = new EventBusService(asJava)
  
  /**
   * Sugar method to creates a record for this type.
   * 
   * The java interface is added to the metadata in the `service.interface` key.   * @param name the name of the service.
   * @param address the event bus address on which the service available
   * @param itf the Java interface (name)
   * @param metadata the metadata
   * @return the created recordsee <a href="../../../../../../../cheatsheet/Record.html">Record</a>
   */
  def createRecord(name: String,address: String,itf: String,metadata: io.vertx.core.json.JsonObject): Record = {
    Record(JEventBusService.createRecord(name.asInstanceOf[java.lang.String], address.asInstanceOf[java.lang.String], itf.asInstanceOf[java.lang.String], metadata))
  }

  /**
   * Lookup for a service record and if found, retrieve it and return the service object (used to consume the service).
   * This is a convenient method to avoid explicit lookup and then retrieval of the service. This method requires to
   * have the `clientClass` set with the expected set of client. This is important for usages not using Java so
   * you can pass the expected type.   * @param discovery the service discovery
   * @param filter the filter
   * @param clientClass the client class
   * @param resultHandler the result handler
   * @return `null` - do not use
   */
  def getServiceProxy[T: TypeTag](discovery: ServiceDiscovery,filter: Record => Boolean,clientClass: Class[T],resultHandler: Handler[AsyncResult[T]]): T = {
    toScala[T](JEventBusService.getServiceProxy[Object](discovery.asJava.asInstanceOf[JServiceDiscovery], {x: JRecord => filter(Record(x)).asInstanceOf[java.lang.Boolean]}, toJavaClass(clientClass), {x: AsyncResult[Object] => resultHandler.handle(AsyncResultWrapper[Object, T](x, a => toScala[T](a)))}))
  }

  /**
   * Lookup for a service record and if found, retrieve it and return the service object (used to consume the service).
   * This is a convenient method to avoid explicit lookup and then retrieval of the service. This method requires to
   * have the `clientClass` set with the expected set of client. This is important for usages not using Java so
   * you can pass the expected type.   * @param discovery the service discovery
   * @param filter the filter as json object
   * @param clientClass the client class
   * @param resultHandler the result handler
   * @return `null` - do not use
   */
  def getServiceProxyWithJsonFilter[T: TypeTag](discovery: ServiceDiscovery,filter: io.vertx.core.json.JsonObject,clientClass: Class[T],resultHandler: Handler[AsyncResult[T]]): T = {
    toScala[T](JEventBusService.getServiceProxyWithJsonFilter[Object](discovery.asJava.asInstanceOf[JServiceDiscovery], filter, toJavaClass(clientClass), {x: AsyncResult[Object] => resultHandler.handle(AsyncResultWrapper[Object, T](x, a => toScala[T](a)))}))
  }

  /**
   * Creates a record based on the parameters.   * @param name the service name
   * @param address the address
   * @param classname the payload class
   * @return the recordsee <a href="../../../../../../../cheatsheet/Record.html">Record</a>
   */
  def createRecord(name: String,address: String,classname: String): Record = {
    Record(JEventBusService.createRecord(name.asInstanceOf[java.lang.String], address.asInstanceOf[java.lang.String], classname.asInstanceOf[java.lang.String]))
  }

}
